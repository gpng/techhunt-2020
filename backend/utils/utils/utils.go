package utils

import (
	"log"
	"mime/multipart"
	"net/http"
	"runtime"

	"golang.org/x/crypto/bcrypt"
)

// ContextKey is the unique key that represents a context value
type ContextKey string

func (c ContextKey) String() string {
	return "context key " + string(c)
}

// HashPassword will generate a bcrypt hash
func HashPassword(password string) (string, error) {
	bytes, err := bcrypt.GenerateFromPassword([]byte(password), 14)
	return string(bytes), err
}

// CheckPasswordHash will compare bcrypt hash with password
func CheckPasswordHash(password, hash string) bool {
	err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
	return err == nil
}

// GetMimeType from file
func GetMimeType(file multipart.File) (string, error) {
	fileHeader := make([]byte, 512)
	if _, err := file.Read(fileHeader); err != nil {
		return "", err
	}

	if _, err := file.Seek(0, 0); err != nil {
		return "", err
	}

	return http.DetectContentType(fileHeader), nil
}

// NewTrue initialises a true bool in memory and returns the address
// for use in struct bool pointers
func NewTrue() *bool {
	b := true
	return &b
}

// IsStringInSlice to check if string present in slice of strings
func IsStringInSlice(str string, list []string) bool {
	for _, s := range list {
		if s == str {
			return true
		}
	}
	return false
}

// LogError to console
func LogError(err error) {
	// get function caller name
	pc, file, line, ok := runtime.Caller(1)
	details := runtime.FuncForPC(pc)
	if ok && details != nil {
		log.Printf("%s:%d:%s failed with error: %v", file, line, details.Name(), err)
	}
}
