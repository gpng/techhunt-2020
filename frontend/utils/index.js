/**
 * Takes a promise, and returns then promised results in format [err, response]
 * @param { Promise } promise
 * @param { Object= } errorExt - Additional Information you can pass to the err object
 * @return { Promise }
 */
export const to = (promise, errorExt) => {
  return promise
    .then((data) => [null, data])
    .catch((err) => {
      if (errorExt) {
        Object.assign(err, errorExt);
      }

      return [err, undefined];
    });
};

export default to;
