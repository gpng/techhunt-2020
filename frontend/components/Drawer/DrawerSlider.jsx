import React, { useState, useRef } from 'react';
import classNames from 'classnames';
// components
import Drawer from './Drawer';
// utils
import { useOnClickOutside } from '../../utils/hooks';

const DrawerSlider = () => {
  const [open, setOpen] = useState(false);

  const containerRef = useRef(null);

  useOnClickOutside(containerRef, () => {
    if (open) {
      setOpen(false);
    }
  });

  return (
    <div className="drawer-slider-root">
      <button type="button" className="button-menu" onClick={() => setOpen(!open)}>
        hamburger
      </button>
      <div
        ref={containerRef}
        className={classNames('drawer-slider-container', {
          'drawer-slider-container--open': open,
        })}
      >
        <Drawer />
      </div>
      <style jsx>{`
        .button-drawer-slider {
        }

        .drawer-slider-container {
          position: fixed;
          height: 100%;
          width: 12rem;
          top: 0;
          left: 0;
          transform: translateX(-12rem);
          transition: all 0.2s;
        }

        .drawer-slider-container--open {
          transform: translateX(0);
        }
      `}</style>
    </div>
  );
};

export default DrawerSlider;
