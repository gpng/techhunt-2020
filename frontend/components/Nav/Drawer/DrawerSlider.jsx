import React, { useState, useRef } from 'react';
import classNames from 'classnames';
// components
import Drawer from './Drawer';
import { Menu } from '../../General/Icons';
// utils
import { useOnClickOutside } from '../../../utils/hooks';
// constants
import { COLORS } from '../../../constants/styles';

const DrawerSlider = () => {
  const [open, setOpen] = useState(false);

  const containerRef = useRef(null);

  useOnClickOutside(containerRef, () => {
    if (open) {
      setOpen(false);
    }
  });

  return (
    <div className="drawer-slider-root">
      <button
        type="button"
        className="button-menu"
        onClick={() => setOpen(!open)}
        data-testid="button-menu"
      >
        <Menu width="1rem" fill={COLORS.BACKGROUND_PRIMARY} />
      </button>
      <div
        ref={containerRef}
        className={classNames('drawer-slider-container', {
          'drawer-slider-container--open': open,
        })}
      >
        <Drawer />
      </div>
      <style jsx>{`
        .button-drawer-slider {
        }

        .button-menu {
          width: 2rem;
          height: 2rem;
          display: flex;
          justify-content: center;
          align-items: center;
          background: transparent;
          border: none;
        }

        .drawer-slider-container {
          position: fixed;
          height: 100%;
          width: 12rem;
          top: 0;
          left: 0;
          transform: translateX(-12rem);
          transition: all 0.2s;
        }

        .drawer-slider-container--open {
          transform: translateX(0);
        }
      `}</style>
    </div>
  );
};

export default DrawerSlider;
