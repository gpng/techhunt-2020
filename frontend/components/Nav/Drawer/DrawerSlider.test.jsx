import React from 'react';
import { cleanup, fireEvent } from '@testing-library/react';
import DrawerSlider from './DrawerSlider';
import { renderWithContext } from '../../../utils/tests';

afterEach(cleanup);

it('should render', () => {
  renderWithContext(<DrawerSlider />);
});

it('should match default snapshot', () => {
  const { asFragment } = renderWithContext(<DrawerSlider />);
  expect(asFragment()).toMatchSnapshot();
});

it('should match open snapshot', () => {
  const { asFragment, getByTestId } = renderWithContext(<DrawerSlider />);
  fireEvent.click(getByTestId('button-menu'));
  expect(asFragment()).toMatchSnapshot();
});

it('should match default snapshot after close', () => {
  const { asFragment, getByTestId } = renderWithContext(<DrawerSlider />);
  fireEvent.click(getByTestId('button-menu'));
  fireEvent.click(getByTestId('button-menu'));
  expect(asFragment()).toMatchSnapshot('should match default snapshot');
});
