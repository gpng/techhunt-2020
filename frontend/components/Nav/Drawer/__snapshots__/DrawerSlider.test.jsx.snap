// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match default snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-2616149830 drawer-slider-root"
  >
    <button
      class="jsx-2616149830 button-menu"
      data-testid="button-menu"
      type="button"
    >
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          fill="#f0f0f0"
        />
      </svg>
    </button>
    <div
      class="jsx-2616149830 drawer-slider-container"
    >
      <aside
        class="jsx-4236861201 drawer-root"
      >
        <div
          class="jsx-961371421 avatar-root"
        >
          <img
            alt=""
            class="jsx-961371421"
            src="/static/images/avatar_default.svg"
          />
        </div>
        <section
          class="jsx-4236861201"
        >
          <p
            class="jsx-4236861201 name"
          >
            Hello, Ashley!
          </p>
        </section>
        <section
          class="jsx-4236861201 locale-container"
        >
          <div
            class="jsx-1468845466 locale-display-root"
          >
            <span
              class="jsx-1468845466"
            >
              Language: English
            </span>
          </div>
        </section>
        <section
          class="jsx-4236861201 button-container"
        >
          <button
            class="jsx-4236861201 button-function"
            data-testid="button-upload"
            type="button"
          >
            Upload CSV
          </button>
        </section>
      </aside>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should match default snapshot after close: should match default snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-2616149830 drawer-slider-root"
  >
    <button
      class="jsx-2616149830 button-menu"
      data-testid="button-menu"
      type="button"
    >
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          fill="#f0f0f0"
        />
      </svg>
    </button>
    <div
      class="jsx-2616149830 drawer-slider-container"
    >
      <aside
        class="jsx-4236861201 drawer-root"
      >
        <div
          class="jsx-961371421 avatar-root"
        >
          <img
            alt=""
            class="jsx-961371421"
            src="/static/images/avatar_default.svg"
          />
        </div>
        <section
          class="jsx-4236861201"
        >
          <p
            class="jsx-4236861201 name"
          >
            Hello, Ashley!
          </p>
        </section>
        <section
          class="jsx-4236861201 locale-container"
        >
          <div
            class="jsx-1468845466 locale-display-root"
          >
            <span
              class="jsx-1468845466"
            >
              Language: English
            </span>
          </div>
        </section>
        <section
          class="jsx-4236861201 button-container"
        >
          <button
            class="jsx-4236861201 button-function"
            data-testid="button-upload"
            type="button"
          >
            Upload CSV
          </button>
        </section>
      </aside>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should match open snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-2616149830 drawer-slider-root"
  >
    <button
      class="jsx-2616149830 button-menu"
      data-testid="button-menu"
      type="button"
    >
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          fill="#f0f0f0"
        />
      </svg>
    </button>
    <div
      class="jsx-2616149830 drawer-slider-container drawer-slider-container--open"
    >
      <aside
        class="jsx-4236861201 drawer-root"
      >
        <div
          class="jsx-961371421 avatar-root"
        >
          <img
            alt=""
            class="jsx-961371421"
            src="/static/images/avatar_default.svg"
          />
        </div>
        <section
          class="jsx-4236861201"
        >
          <p
            class="jsx-4236861201 name"
          >
            Hello, Ashley!
          </p>
        </section>
        <section
          class="jsx-4236861201 locale-container"
        >
          <div
            class="jsx-1468845466 locale-display-root"
          >
            <span
              class="jsx-1468845466"
            >
              Language: English
            </span>
          </div>
        </section>
        <section
          class="jsx-4236861201 button-container"
        >
          <button
            class="jsx-4236861201 button-function"
            data-testid="button-upload"
            type="button"
          >
            Upload CSV
          </button>
        </section>
      </aside>
    </div>
  </div>
</DocumentFragment>
`;
